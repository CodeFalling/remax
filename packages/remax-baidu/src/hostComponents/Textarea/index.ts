import * as React from 'react';
import { createHostComponent } from '@remax/shared';
import { BaseProps, GenericEvent } from '../../types/component';

export interface TextareaProps extends BaseProps {
  value?: string;
  disabled?: boolean;
  maxLength?: number;
  placeholder?: string;
  placeholderStyle?: React.CSSProperties;
  placeholderClassName?: string;
  autoHeight?: boolean;
  cursor?: number;
  autoFocus?: boolean;
  confirmType?: 'default' | 'done' | 'send' | 'search' | 'next' | 'go';
  confirmHold?: boolean;
  focus?: boolean;
  fixed?: boolean;
  cursorSpacing?: number;
  showConfirmBar?: boolean;
  selectionStart?: number;
  selectionEnd?: number;
  adjustPosition?: boolean;
  onFocus?: (e: GenericEvent) => void;
  onBlur?: (e: GenericEvent) => void;
  onLinechange?: (e: GenericEvent) => void;
  onInput?: (e: GenericEvent) => void;
  onConfirm?: (e: GenericEvent) => void;
}

export const Textarea: React.ComponentType<TextareaProps> = createHostComponent<TextareaProps>('textarea');

Textarea.defaultProps = {
  disabled: false,
  maxLength: 140,
  autoHeight: false,
  cursor: -1,
  autoFocus: false,
  confirmType: 'default',
  confirmHold: false,
  focus: false,
  fixed: false,
  cursorSpacing: 0,
  showConfirmBar: true,
  selectionStart: -1,
  selectionEnd: -1,
  adjustPosition: true,
};
