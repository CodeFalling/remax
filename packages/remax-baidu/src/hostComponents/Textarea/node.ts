export const alias = {
  id: 'id',
  className: 'class',
  style: 'style',
  value: 'value',
  disabled: 'disabled',
  maxLength: 'maxlength',
  placeholder: 'placeholder',
  placeholderStyle: 'placeholder-style',
  placeholderClassName: 'placeholder-class',
  autoHeight: 'auto-height',
  cursor: 'cursor',
  autoFocus: 'auto-focus',
  confirmType: 'confirm-type',
  confirmHold: 'confirm-hold',
  focus: 'focus',
  fixed: 'fixed',
  cursorSpacing: 'cursor-spacing',
  showConfirmBar: 'show-confirm-bar',
  selectionStart: 'selection-start',
  selectionEnd: 'selection-end',
  adjustPosition: 'adjust-position',
  onFocus: 'bindfocus',
  onBlur: 'bindblur',
  onLinechange: 'bindlinechange',
  onInput: 'bindinput',
  onConfirm: 'bindconfirm',
};

export const props = Object.values(alias);
