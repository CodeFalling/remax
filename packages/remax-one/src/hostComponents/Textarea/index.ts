import * as React from 'react';
import { InputEvent } from '../../types';
import { createCallback, createInputEvent } from '../../createHostComponent';

export interface TextareaProps {
  // 通用属性
  readonly dataset?: DOMStringMap;
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  name?: string;
  value?: any;
  defaultValue?: any;
  placeholder?: string;
  placeholderStyle?: React.CSSProperties;
  disabled?: boolean;
  /** 最大输入长度，设置为 -1 的时候不限制最大长度 */
  maxlength?: number;
  focus?: boolean;
  /** 是否自动增高，设置auto-height时，style.height不生效 */
  autoHeight?: boolean;
  onInput?: (event: InputEvent) => any;
  onFocus?: (event: InputEvent) => void;
  onBlur?: (event: InputEvent) => void;
  onConfirm?: (event: InputEvent) => void;
}

export interface TextareaState {
  value?: string;
  controlled: boolean;
}

export default class Textarea extends React.Component<TextareaProps, TextareaState> {
  static defaultProps = {
    'toutiao-maxlength': -1,
    'toutiao-selection-end': 999,
    'toutiao-selection-start': 999,
    'wechat-maxlength': -1,
    'wechat-selection-end': 999,
    'wechat-selection-start': 999,
    'wechat-fixed': false,
  };

  state: TextareaState = {
    value: '',
    // 支付宝
    controlled: false,
  };

  constructor(props: TextareaProps) {
    super(props);

    const controlled = props.value !== undefined;
    const value = controlled ? props.value : props.defaultValue;

    this.state = {
      value,
      controlled,
    };
  }

  componentDidUpdate() {
    if (this.props.value !== undefined && this.props.value !== this.state.value) {
      this.setState({ value: this.props.value });
    }
  }

  handleInput = (e: InputEvent) => {
    const { controlled } = this.state;

    if (!controlled) {
      this.setState({ value: e.target.value });
    }

    if (typeof this.props.onInput === 'function') {
      return this.props.onInput(e);
    }

    // 微信
    // 注意，微信要对 input 受控，必须要在 onInput 方法返回值
    return controlled ? this.props.value : e.target.value;
  };

  render() {
    const inputProps = {
      ...this.props,
      onInput: createCallback(this.handleInput, createInputEvent),
    };

    if (inputProps.onConfirm) {
      inputProps.onConfirm = createCallback(this.props.onConfirm, createInputEvent);
    }

    if (inputProps.onFocus) {
      inputProps.onFocus = createCallback(this.props.onFocus, createInputEvent);
    }

    if (inputProps.onBlur) {
      inputProps.onBlur = createCallback(this.props.onBlur, createInputEvent);
    }

    return React.createElement('textarea', {
      ...inputProps,
      ...this.state,
    });
  }
}
