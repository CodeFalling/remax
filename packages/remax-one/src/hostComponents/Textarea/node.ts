const platformAlias: any = {
  alipay: {
    onInput: 'onInput',
    onFocus: 'onFocus',
    onBlur: 'onBlur',
    onConfirm: 'onConfirm',
  },
  wechat: {
    onInput: 'bindinput',
    onFocus: 'bindfocus',
    onBlur: 'bindblur',
    onConfirm: 'bindconfirm',
  },
  toutiao: {
    onInput: 'bindinput',
    onFocus: 'bindfocus',
    onBlur: 'bindblur',
    onConfirm: 'bindconfirm',
  },
};

const defaultAlias = {
  id: 'id',
  className: 'class',
  style: 'style',
  name: 'name',
  value: 'value',
  placeholder: 'placeholder',
  placeholderStyle: 'placeholder-style',
  disabled: 'disabled',
  maxlength: 'maxlength',
  focus: 'focus',
  autoHeight: 'auto-height',
  onInput: 'onInput',
  onFocus: 'onFocus',
  onBlur: 'onBlur',
  onConfirm: 'onConfirm',
};

export const alias = { ...defaultAlias, ...platformAlias[process.env.REMAX_PLATFORM!] };

export const props = Object.values(alias);
