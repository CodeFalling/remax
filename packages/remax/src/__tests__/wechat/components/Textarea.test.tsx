import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import Textarea from '../../../adapters/wechat/components/Textarea';

describe('TextArea', () => {
  it('render correctly', () => {
    const testRenderer = TestRenderer.create(
      <Textarea className="class" value="" />
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
