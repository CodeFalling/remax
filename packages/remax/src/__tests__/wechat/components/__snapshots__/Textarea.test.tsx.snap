// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea render correctly 1`] = `
<textarea
  bindblur={[Function]}
  bindfocus={[Function]}
  bindinput={[Function]}
  bindtap={[Function]}
  class="class"
  fixed={false}
  focus={false}
  maxlength={-1}
  selectionEnd={999}
  selectionStart={999}
  value=""
/>
`;
